TextEditor.h

```diff
@@ -182,6 +182,60 @@ class TextEditor
 		static const LanguageDefinition& Lua();
 	};
 
+	struct EditorState
+	{
+		Coordinates mSelectionStart;
+		Coordinates mSelectionEnd;
+		Coordinates mCursorPosition;
+	};
+
+	class UndoRecord
+	{
+	public:
+		UndoRecord() {}
+		~UndoRecord() {}
+
+		UndoRecord(
+			const std::string& aAdded,
+			const TextEditor::Coordinates aAddedStart,
+			const TextEditor::Coordinates aAddedEnd,
+
+			const std::string& aRemoved,
+			const TextEditor::Coordinates aRemovedStart,
+			const TextEditor::Coordinates aRemovedEnd,
+
+			TextEditor::EditorState& aBefore,
+			TextEditor::EditorState& aAfter);
+
+		UndoRecord(UndoRecord& undoRecord);
+
+		void Undo(TextEditor* aEditor);
+		void Redo(TextEditor* aEditor);
+
+		std::string mAdded;
+		Coordinates mAddedStart;
+		Coordinates mAddedEnd;
+
+		std::string mRemoved;
+		Coordinates mRemovedStart;
+		Coordinates mRemovedEnd;
+
+		EditorState mBefore;
+		EditorState mAfter;
+	};
+
+	class ExternalUndoBufferInterface
+	{
+	public:
+		ExternalUndoBufferInterface() {};
+		~ExternalUndoBufferInterface() {};
+
+		virtual void AddUndo(UndoRecord&, TextEditor&)=0;
+

```

Comment:
Maybe I should pass the text editor ref in the constructor instead of in the AddUndo method.